### ISSUE TITLE
parallelization issues

### ISSUE THREAD

**Issue by beauraF** [titusfortner/webdrivers#77]
Hello,

_Wedrivers_ is having some problems with parallelization. This makes it unusable on _Heroku CI_ and under _Rails 6_.
One of the threads downloads the driver while the others try to access the resource. Which leads to this error.

```
Error:
AuthenticationsTest#test_successful_login:
Zip::Error: File /home/circleci/.webdrivers/chromedriver_linux64.zip has zero size. Did you mean to pass the create flag?
    test/system/authentications_test.rb:7:in `block in <class:AuthenticationsTest>'


rails test test/system/authentications_test.rb:6
```

```
Error:
AuthenticationsTest#test_failed_login:
ChildProcess::LaunchError: Text file busy - /home/circleci/.webdrivers/chromedriver
    test/system/authentications_test.rb:14:in `block in <class:AuthenticationsTest>'


rails test test/system/authentications_test.rb:13

Traceback (most recent call last):
	13: from /home/circleci/doctopod/vendor/bundle/gems/minitest-5.11.3/lib/minitest.rb:63:in `block in autorun'
	12: from /home/circleci/doctopod/vendor/bundle/gems/minitest-5.11.3/lib/minitest.rb:133:in `run'
	11: from /home/circleci/doctopod/vendor/bundle/gems/activesupport-6.0.0.rc1/lib/active_support/testing/parallelization.rb:70:in `start'
	10: from /home/circleci/doctopod/vendor/bundle/gems/activesupport-6.0.0.rc1/lib/active_support/testing/parallelization.rb:70:in `map'
	 9: from /home/circleci/doctopod/vendor/bundle/gems/activesupport-6.0.0.rc1/lib/active_support/testing/parallelization.rb:70:in `each'
	 8: from /home/circleci/doctopod/vendor/bundle/gems/activesupport-6.0.0.rc1/lib/active_support/testing/parallelization.rb:70:in `times'
	 7: from /home/circleci/doctopod/vendor/bundle/gems/activesupport-6.0.0.rc1/lib/active_support/testing/parallelization.rb:71:in `block in start'
	 6: from /home/circleci/doctopod/vendor/bundle/gems/activesupport-6.0.0.rc1/lib/active_support/testing/parallelization.rb:71:in `fork'
	 5: from /home/circleci/doctopod/vendor/bundle/gems/selenium-webdriver-3.142.0/lib/selenium/webdriver/common/platform.rb:150:in `block in exit_hook'
	 4: from /home/circleci/doctopod/vendor/bundle/gems/selenium-webdriver-3.142.0/lib/selenium/webdriver/common/service.rb:105:in `stop'
	 3: from /home/circleci/doctopod/vendor/bundle/gems/selenium-webdriver-3.142.0/lib/selenium/webdriver/common/service.rb:170:in `stop_server'
	 2: from /home/circleci/doctopod/vendor/bundle/gems/selenium-webdriver-3.142.0/lib/selenium/webdriver/common/service.rb:180:in `process_exited?'
	 1: from /home/circleci/doctopod/vendor/bundle/gems/childprocess-1.0.1/lib/childprocess/unix/process.rb:31:in `exited?'
/home/circleci/doctopod/vendor/bundle/gems/childprocess-1.0.1/lib/childprocess/abstract_process.rb:188:in `assert_started': process not started (ChildProcess::Error)
	14: from /home/circleci/doctopod/vendor/bundle/gems/minitest-5.11.3/lib/minitest.rb:63:in `block in autorun'
	13: from /home/circleci/doctopod/vendor/bundle/gems/minitest-5.11.3/lib/minitest.rb:133:in `run'
	12: from /home/circleci/doctopod/vendor/bundle/gems/activesupport-6.0.0.rc1/lib/active_support/testing/parallelization.rb:70:in `start'
	11: from /home/circleci/doctopod/vendor/bundle/gems/activesupport-6.0.0.rc1/lib/active_support/testing/parallelization.rb:70:in `map'
	10: from /home/circleci/doctopod/vendor/bundle/gems/activesupport-6.0.0.rc1/lib/active_support/testing/parallelization.rb:70:in `each'
	 9: from /home/circleci/doctopod/vendor/bundle/gems/activesupport-6.0.0.rc1/lib/active_support/testing/parallelization.rb:70:in `times'
	 8: from /home/circleci/doctopod/vendor/bundle/gems/activesupport-6.0.0.rc1/lib/active_support/testing/parallelization.rb:71:in `block in start'
	 7: from /home/circleci/doctopod/vendor/bundle/gems/activesupport-6.0.0.rc1/lib/active_support/testing/parallelization.rb:71:in `fork'
	 6: from /home/circleci/doctopod/vendor/bundle/gems/selenium-webdriver-3.142.0/lib/selenium/webdriver/common/platform.rb:150:in `block in exit_hook'
	 5: from /home/circleci/doctopod/vendor/bundle/gems/selenium-webdriver-3.142.0/lib/selenium/webdriver/common/service.rb:110:in `stop'
	 4: from /home/circleci/doctopod/vendor/bundle/gems/selenium-webdriver-3.142.0/lib/selenium/webdriver/common/service.rb:110:in `ensure in stop'
	 3: from /home/circleci/doctopod/vendor/bundle/gems/selenium-webdriver-3.142.0/lib/selenium/webdriver/common/service.rb:163:in `stop_process'
	 2: from /home/circleci/doctopod/vendor/bundle/gems/selenium-webdriver-3.142.0/lib/selenium/webdriver/common/service.rb:180:in `process_exited?'
	 1: from /home/circleci/doctopod/vendor/bundle/gems/childprocess-1.0.1/lib/childprocess/unix/process.rb:31:in `exited?'
/home/circleci/doctopod/vendor/bundle/gems/childprocess-1.0.1/lib/childprocess/abstract_process.rb:188:in `assert_started': process not started (ChildProcess::Error)
```

These are logs under Rails 6 and Circle CI. But I get the same result on Rails 5.2.3 with Heroku CI.

If you have any idea how we can handle this, I'm ready to work on it.

Thanks! Keep up the good work!

**Comment by Govan** [comment]
@beauraF When running under parallel_test we're able to solve this by downloading one driver per process....

```
require 'webdrivers'
Webdrivers::Chromedriver.version = '2.35'
Webdrivers.install_dir = File.expand_path('~/.webdrivers/' + ENV['TEST_ENV_NUMBER'].to_s)

```

**Comment by kevindew** [comment]
I've just experienced the same problem. We have a Rails app that is started with foreman with web and worker processes. These hit the same empty zip file issue.

Some ideas to work around this could be:

1. Giving the zip a UUID based file name (or dealing with it as a temp file) so two processes can't be
2. Keeping the zip around afterwards rather than deleting it, so a second process decompressing it isn't fatal
3. Less nice, but could have a rake task that could be run before parallelisation to initialise webdrivers

**Comment by titusfortner** [comment]
Rake tasks are a good idea, regardless. I'll see if I can figure out how to add those intelligently.

As for collisions, I don't want to have someone running 30 tests in parallel that each download a new driver, regardless of how we would keep them from overwriting each other. I'm thinking maybe we can create a local file and use its timestamp to "lock" new downloads. It's also how I'm planning on implementing #29 so can probably pull double duty. We'll see what we can do.

**Comment by titusfortner** [comment]
just merged #84 which will address the immediate issue if not the underlying problem.

**Comment by beauraF** [comment]
Awesome. It's a first step. Are you planning a release soon?

**Comment by titusfortner** [comment]
Yes, I'm loading up my old Windows machine to validate everything is working there first.

**Comment by titusfortner** [comment]
@beauraF @Govan @kevindew can you try 3.9 and set a cache value (recommended `Webdrivers.cache_time = 86_400`) and let me know if this is still an issue?

**Comment by beauraF** [comment]
Same issue on my side: https://circleci.com/gh/beauraF/doctopod/57

**Comment by titusfortner** [comment]
Oh, I think I see what is happening. I'll get a point release with the change and we can reevaluate.

**Comment by beauraF** [comment]
@titusfortner If you need access to my project/my CI to perform tests. No worries, I can provide you an access.

**Comment by ryan-plated** [comment]
I just run `Webdrivers::Chromedriver.update` before fork

**Comment by titusfortner** [comment]
Yeah, I think that's going to need to be the recommended solution, but we'll create some rake tasks in 4.0 to give people the option of doing it from command line instead of forcing everyone to execute it in their code.

**Comment by titusfortner** [comment]
Rake tasks have been merged with #117

**Comment by kevindew** [comment]
Thanks for your work on this @titusfortner 👍

**Comment by mike1o1** [comment]
It wasn't quite clear to me what the fix was, so I'm just adding this here incase others have the same issue.

To fix this with, for example, Circle CI, I added a new run task before my system tests run to execute the rake tasks mentioned above to update the drivers.

```yaml
- run:
    name: Run tests
    environment:
      - CIRCLE_TEST_REPORTS: test/reports/test
    command: bundle exec rails test
    when: always

# This is the new task, run before the system tests execute
- run:
    name: Extract browser
    command: bundle exec rake webdrivers:chromedriver:update
    when: always

- run:
    name: Run system tests
    environment:
      - CIRCLE_TEST_REPORTS: test/reports/system
    command: bundle exec rails test:system
    when: always

```

**Comment by map7** [comment]
I'm still on Rails 5.2.3 using the parallel_tests gem and ran into the same problem. The only fix which worked for me is;

    Webdrivers.install_dir = File.expand_path('~/.webdrivers/' + ENV['TEST_ENV_NUMBER'].to_s)

**Comment by kevx86** [comment]
> Webdrivers.install_dir = File.expand_path('~/.webdrivers/' + ENV['TEST_ENV_NUMBER'].to_s)

@map7 Where did you place this line? Thanks,

**Comment by map7** [comment]
@kevx86 I put it in the specs/rails_helper.rb and set environment variables in my .bashrc/.zshrc file
